{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23387808-3c4b-4787-97d1-36cba5a1ed6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e47d083-124d-415d-a013-576987305d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading data DataFrame\n",
    "df_hocomoco = pd.read_csv('H12_complete_dataframe.csv')\n",
    "df_jaspar = pd.read_csv('JASPAR2024_CORE_all_taxonomy.csv')\n",
    "df_animaltfdb = pd.read_csv('New_matrix_AnimalRFDB.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9ed2e861-0946-4986-b456-dcc4d4b0afd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SHAPE DataFrame Hocomoco: (1443, 19)\n",
      "\n",
      "SHAPE DataFrame Jaspar: (1362, 8)\n",
      "\n",
      "SHAPE DataFrame AnimalTFDB: (404, 4)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for idx, df in enumerate([df_hocomoco, df_jaspar, df_animaltfdb]):\n",
    "    print(f'SHAPE DataFrame {['Hocomoco', 'Jaspar', 'AnimalTFDB'][idx]}: {df.shape}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e7a1ab-5ec4-48fa-9147-38197a5fb5fc",
   "metadata": {},
   "source": [
    "### Merge dataframes into one dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f50bc5ce-e06e-411e-b2ea-e2e1da76c6d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Motif</th>\n",
       "      <th>LOGO</th>\n",
       "      <th>Gene (human)</th>\n",
       "      <th>Gene (mouse)</th>\n",
       "      <th>Motif length</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Motif subtype</th>\n",
       "      <th>Consensus</th>\n",
       "      <th>Quality.1</th>\n",
       "      <th>TF family</th>\n",
       "      <th>TF subfamily</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>UniProt ID (human)</th>\n",
       "      <th>UniProt ID (mouse)</th>\n",
       "      <th>path_pcm</th>\n",
       "      <th>path_pwm</th>\n",
       "      <th>path_pfm</th>\n",
       "      <th>seqs_HUMAN</th>\n",
       "      <th>seqs_MOUSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AHR.H12CORE.0.P.B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AHR</td>\n",
       "      <td>Ahr</td>\n",
       "      <td>10</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>vhbWGCGTGM</td>\n",
       "      <td>P</td>\n",
       "      <td>PAS {1.2.5}</td>\n",
       "      <td>PAS-AHR {1.2.5.1}</td>\n",
       "      <td>HGNC:348</td>\n",
       "      <td>AHR_HUMAN</td>\n",
       "      <td>AHR_MOUSE</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/AHR.H12CORE...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/AHR.H12CORE...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/AHR.H12CORE...</td>\n",
       "      <td>&gt;sp|P35869|AHR_HUMAN Aryl hydrocarbon receptor...</td>\n",
       "      <td>&gt;sp|P30561|AHR_MOUSE Aryl hydrocarbon receptor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AHRR.H12CORE.0.P.C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AHRR</td>\n",
       "      <td>Ahrr</td>\n",
       "      <td>11</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>bYKCGTGMvWd</td>\n",
       "      <td>P</td>\n",
       "      <td>PAS {1.2.5}</td>\n",
       "      <td>PAS-AHR {1.2.5.1}</td>\n",
       "      <td>HGNC:346</td>\n",
       "      <td>AHRR_HUMAN</td>\n",
       "      <td>AHRR_MOUSE</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/AHRR.H12COR...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/AHRR.H12COR...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/AHRR.H12COR...</td>\n",
       "      <td>&gt;sp|A9YTQ3|AHRR_HUMAN Aryl hydrocarbon recepto...</td>\n",
       "      <td>&gt;sp|Q3U1U7|AHRR_MOUSE Aryl hydrocarbon recepto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALX1.H12CORE.0.SM.B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ALX1</td>\n",
       "      <td>Alx1</td>\n",
       "      <td>20</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>nvbTAATTRRRTTAvnnnnn</td>\n",
       "      <td>S+M</td>\n",
       "      <td>Paired-related HD {3.1.3}</td>\n",
       "      <td>ALX {3.1.3.1}</td>\n",
       "      <td>HGNC:1494</td>\n",
       "      <td>ALX1_HUMAN</td>\n",
       "      <td>ALX1_MOUSE</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/ALX1.H12COR...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/ALX1.H12COR...</td>\n",
       "      <td>Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/ALX1.H12COR...</td>\n",
       "      <td>&gt;sp|Q15699|ALX1_HUMAN ALX homeobox protein 1 O...</td>\n",
       "      <td>&gt;sp|Q8C8B0|ALX1_MOUSE ALX homeobox protein 1 O...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Motif  LOGO Gene (human) Gene (mouse)  Motif length Quality  \\\n",
       "0    AHR.H12CORE.0.P.B   NaN          AHR          Ahr            10       B   \n",
       "1   AHRR.H12CORE.0.P.C   NaN         AHRR         Ahrr            11       C   \n",
       "2  ALX1.H12CORE.0.SM.B   NaN         ALX1         Alx1            20       B   \n",
       "\n",
       "   Motif subtype             Consensus Quality.1                  TF family  \\\n",
       "0              0            vhbWGCGTGM         P                PAS {1.2.5}   \n",
       "1              0           bYKCGTGMvWd         P                PAS {1.2.5}   \n",
       "2              0  nvbTAATTRRRTTAvnnnnn       S+M  Paired-related HD {3.1.3}   \n",
       "\n",
       "        TF subfamily       HGNC UniProt ID (human) UniProt ID (mouse)  \\\n",
       "0  PAS-AHR {1.2.5.1}   HGNC:348          AHR_HUMAN          AHR_MOUSE   \n",
       "1  PAS-AHR {1.2.5.1}   HGNC:346         AHRR_HUMAN         AHRR_MOUSE   \n",
       "2      ALX {3.1.3.1}  HGNC:1494         ALX1_HUMAN         ALX1_MOUSE   \n",
       "\n",
       "                                            path_pcm  \\\n",
       "0  Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/AHR.H12CORE...   \n",
       "1  Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/AHRR.H12COR...   \n",
       "2  Эукариоты/HOCOMOCO/H12CORE_pcm/pcm/ALX1.H12COR...   \n",
       "\n",
       "                                            path_pwm  \\\n",
       "0  Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/AHR.H12CORE...   \n",
       "1  Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/AHRR.H12COR...   \n",
       "2  Эукариоты/HOCOMOCO/H12CORE_pwm/pwm/ALX1.H12COR...   \n",
       "\n",
       "                                            path_pfm  \\\n",
       "0  Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/AHR.H12CORE...   \n",
       "1  Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/AHRR.H12COR...   \n",
       "2  Эукариоты/HOCOMOCO/H12CORE_pfm/pfm/ALX1.H12COR...   \n",
       "\n",
       "                                          seqs_HUMAN  \\\n",
       "0  >sp|P35869|AHR_HUMAN Aryl hydrocarbon receptor...   \n",
       "1  >sp|A9YTQ3|AHRR_HUMAN Aryl hydrocarbon recepto...   \n",
       "2  >sp|Q15699|ALX1_HUMAN ALX homeobox protein 1 O...   \n",
       "\n",
       "                                          seqs_MOUSE  \n",
       "0  >sp|P30561|AHR_MOUSE Aryl hydrocarbon receptor...  \n",
       "1  >sp|Q3U1U7|AHRR_MOUSE Aryl hydrocarbon recepto...  \n",
       "2  >sp|Q8C8B0|ALX1_MOUSE ALX homeobox protein 1 O...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hocomoco.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e4651716-d5cb-481a-aa23-8588b27b8f42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Matrix</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>TF_family</th>\n",
       "      <th>TF_class</th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>Data_type</th>\n",
       "      <th>Sequence_TF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MA0010.2</td>\n",
       "      <td>[[1.0, 1.0, 1.0, 6.0], [5.0, 2.0, 1.0, 1.0], [...</td>\n",
       "      <td>insects</td>\n",
       "      <td>Other factors with up to three adjacent zinc f...</td>\n",
       "      <td>C2H2 zinc finger factors</td>\n",
       "      <td>Q01295</td>\n",
       "      <td>COMPILED</td>\n",
       "      <td>&gt;sp|Q01295|BRC1_DROME Broad-complex core prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0011.2</td>\n",
       "      <td>[[0.0, 10.0, 0.0, 2.0], [0.0, 1.0, 0.0, 11.0],...</td>\n",
       "      <td>insects</td>\n",
       "      <td>Other factors with up to three adjacent zinc f...</td>\n",
       "      <td>C2H2 zinc finger factors</td>\n",
       "      <td>Q01295</td>\n",
       "      <td>COMPILED</td>\n",
       "      <td>&gt;sp|Q01295|BRC1_DROME Broad-complex core prote...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID                                             Matrix Taxonomy  \\\n",
       "0  MA0010.2  [[1.0, 1.0, 1.0, 6.0], [5.0, 2.0, 1.0, 1.0], [...  insects   \n",
       "1  MA0011.2  [[0.0, 10.0, 0.0, 2.0], [0.0, 1.0, 0.0, 11.0],...  insects   \n",
       "\n",
       "                                           TF_family  \\\n",
       "0  Other factors with up to three adjacent zinc f...   \n",
       "1  Other factors with up to three adjacent zinc f...   \n",
       "\n",
       "                   TF_class Uniprot Data_type  \\\n",
       "0  C2H2 zinc finger factors  Q01295  COMPILED   \n",
       "1  C2H2 zinc finger factors  Q01295  COMPILED   \n",
       "\n",
       "                                         Sequence_TF  \n",
       "0  >sp|Q01295|BRC1_DROME Broad-complex core prote...  \n",
       "1  >sp|Q01295|BRC1_DROME Broad-complex core prote...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_jaspar.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3fe24e34-58e3-450d-9b50-a053f5c3c4fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Matrix</th>\n",
       "      <th>Protein_sequence</th>\n",
       "      <th>Tresh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V_AHR_Q6</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0], [0.958333, 0.041667, 0....</td>\n",
       "      <td>&gt;sp|A9YTQ3|AHRR_HUMAN Aryl hydrocarbon recepto...</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>V_ALX1_05</td>\n",
       "      <td>[[0.138864, 0.266416, 0.123031, 0.471688], [0....</td>\n",
       "      <td>&gt;sp|Q15699|ALX1_HUMAN ALX homeobox protein 1 O...</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID                                             Matrix  \\\n",
       "0   V_AHR_Q6  [[0.0, 1.0, 0.0, 0.0], [0.958333, 0.041667, 0....   \n",
       "1  V_ALX1_05  [[0.138864, 0.266416, 0.123031, 0.471688], [0....   \n",
       "\n",
       "                                    Protein_sequence  Tresh  \n",
       "0  >sp|A9YTQ3|AHRR_HUMAN Aryl hydrocarbon recepto...     33  \n",
       "1  >sp|Q15699|ALX1_HUMAN ALX homeobox protein 1 O...     56  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_animaltfdb.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5072a83e-6420-4f33-a6ec-9089c522983f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_file_in_matrix(path_file: list) -> np.array:\n",
    "    \"\"\"Loading a matrix from a file in the format np.array \n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    path_file : the path to the file\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    matrix : the matrix in the format np.array\n",
    "    \n",
    "    \"\"\"\n",
    "    with open(f'../../{path_file}', 'r', encoding='utf-8') as f:\n",
    "        text = f.readlines()\n",
    "\n",
    "    text_matrix = text[1:]\n",
    "    matrix = np.zeros((len(text_matrix), 4))\n",
    "    for i in range(len(text_matrix)):\n",
    "        matrix[i] = list(map(float, text_matrix[i].split('\\t')))\n",
    "\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c86a344-ecc7-4152-b9da-729c95a2ae03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_motif_databases(\n",
    "    hocomoco: pd.DataFrame,\n",
    "    jaspar: pd.DataFrame,\n",
    "    animal: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Combines data from three motive databases into a single DataFrame\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    hocomoco : table with data from database HOCOMOCO (pd.DataFrame)\n",
    "    jaspar : table with data from database Japspar (pd.DataFrame)\n",
    "    animal : table with data from database AnimalTFDB (pd.DataFrame)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    merge_df : a dataframe with combined data from three databases; \\ \n",
    "               columns: ids, name_database, matrix, uniprot_id \n",
    "    \"\"\"\n",
    "    \n",
    "    # Combining ids\n",
    "    motif_ids = (\n",
    "        hocomoco['Motif'].tolist() +\n",
    "        jaspar['ID'].tolist() +\n",
    "        animal['ID'].tolist())\n",
    "\n",
    "    # Matrix processing\n",
    "    def process_hocomoco_matrices(df: pd.DataFrame) -> list:\n",
    "        \"\"\"Converting HOCOMOCO files to matrices\"\"\"\n",
    "        return [convert_file_in_matrix(path) for path in df.path_pfm]\n",
    "\n",
    "    def process_jaspar_matrices(df: pd.DataFrame) -> list:\n",
    "        \"\"\"Normalization of JASPAR matrices\"\"\"\n",
    "        return [\n",
    "            normalize_matrix(np.array(eval(matrix)))\n",
    "            for matrix in df.Matrix.tolist()]\n",
    "\n",
    "    def normalize_matrix(matrix: np.ndarray) -> np.ndarray:\n",
    "        \"\"\"Normalization of the matrix by columns\"\"\"\n",
    "        return matrix / matrix.sum(axis=1, keepdims=True)\n",
    "\n",
    "    matrices = (\n",
    "        process_hocomoco_matrices(hocomoco) +\n",
    "        process_jaspar_matrices(jaspar) +\n",
    "        list(map(lambda x: np.array(eval(x)), animal.Matrix.tolist()))\n",
    "    )\n",
    "\n",
    "    # UniProt ID processing\n",
    "    def extract_uniprot(sequence: str) -> str:\n",
    "        \"\"\"Extracting a UniProt ID from a sequence\"\"\"\n",
    "        try:\n",
    "            return '_'.join(sequence.split(' ')[0].split('|')[1:])\n",
    "        except:\n",
    "            return None\n",
    "\n",
    "    uniprot_ids = (\n",
    "        [extract_uniprot(seq) for seq in hocomoco[\"seqs_HUMAN\"]]\n",
    "        + [extract_uniprot(seq) for seq in jaspar[\"Sequence_TF\"]]\n",
    "        + [extract_uniprot(seq) for seq in animal[\"Protein_sequence\"]]\n",
    "    )\n",
    "\n",
    "    # Creating the resulting DataFrame\n",
    "    data = {\n",
    "        'ID': motif_ids,\n",
    "        'DATABASE': (\n",
    "            ['HOCOMOCO'] * len(hocomoco) +\n",
    "            ['JASPAR'] * len(jaspar) +\n",
    "            ['AnimalTFDB'] * len(animal)\n",
    "        ),\n",
    "        'MATRIX': list(map(lambda x: x.tolist(), matrices)),\n",
    "        'PROTEIN': uniprot_ids\n",
    "    }\n",
    "\n",
    "    merged_df = pd.DataFrame(data).dropna()\n",
    "    return merged_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e1e5adcc-614f-47ab-bc02-a59bf1e7377d",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_df = merge_motif_databases(df_hocomoco, df_jaspar, df_animaltfdb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aa5b10d5-afe7-4d2c-aefa-502d5419b587",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_df = merge_df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59e67627-3a53-46ed-a683-a1fbd87f05de",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_df.to_csv('dataset_matrices.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
